<a href="https://colab.research.google.com/github/mohitsshetty986/Movie-Recommendation-Systems/blob/master/Simple_Recommenders.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***Simple Recommenders***

## **simple recommenders are basic systems that recommend the top items based on a certain metric or score.**
**For example the top 10 imdb rated movies or top 10 netflix rated movies**

###We are getting the top 20 best rated movies of all time using simple recommenders.

**1) To load your dataset, you would be using the pandas DataFrame library.**

In [ ]:
import pandas as pd
import io

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving movies_metadata.csv to movies_metadata.csv


In [ ]:
metadata=pd.read_csv('movies_metadata.csv',low_memory=False)

In [ ]:
metadata.head()

adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0
2  False  ...       92.0
3  False  ...       34.0
4  False  ...      173.0

[5 rows x 24 columns]

**2) Movie Ratings (vote_average from the dataframe above) could be used as a metric to recommend movies**

####But we cannot take the vote_average directly from the dataframe to decide the top recommendations for the user because it has few caveats.

####The vote_average is the average rating of no of users who have voted for the movies. Now we need to consider this scenario that a movie can be rated by '10' voters and get a rating of '9' and, another movie is rated by '1000' voters and get a rating of '8.5'. 

####Now if we had taken the vote_average as it is from the dataframe then, the movie with '10' votes and '9' rating would have been recommended to the user instead of a movie with '1000' votes and '8.5' ratings which clearly shows the latter to be better movie than the previous one.



**3) Taking these shortcomings into consideration, you must come up with a weighted rating that takes into account the average rating and the number of votes it has accumulated.**

**WeightedRating(WR)=(v.R/(v+m))+(m.C/(v+m))**

**This formula will be used to filter out the movies with a certain minimum no of votes and only these movies will be considered for calculating the top 10 or top 20 movies**

---



where,
  v is the number of votes for the movie  ---- vote_count in this case

  m is the minimum votes required to be listed in the chart

  R is the average rating of the movie   --- vote_average in this case

  C is the mean vote across the whole report

In [ ]:
C = metadata['vote_average'].mean()
print(C)

5.618207215133889


**4)For this example I'm considering 90 percentile as the cutoff value i.e. a movie should have votes more than 90% of all the movies in the dataset.**

In [ ]:
m = metadata['vote_count'].quantile(0.90)
print(m)

160.0


**5)From above we find that a movie must have atleast 160 votes to be compared for the top 20 best alltime movies**

In [ ]:
fraction_movies = metadata.copy().loc[metadata['vote_count'] >= m]

In [ ]:
fraction_movies.shape

(4555, 24)

**6) Now lets calculate the weighted ratings for the movies in the updated dataset(movies with votes more than or equal to 160) to get the ratings which we are concerned with. **

In [ ]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
fraction_movies['score'] = fraction_movies.apply(weighted_rating, axis=1)

In [ ]:
fraction_movies.head()

adult  ...     score
0  False  ...  7.640253
1  False  ...  6.820293
4  False  ...  5.660700
5  False  ...  7.537201
8  False  ...  5.556626

[5 rows x 25 columns]

**7) Sort the movies list in descending order to get the top rated movies at the top of the dataframe.**

In [ ]:
fraction_movies = fraction_movies.sort_values('score', ascending=False)

**8) The final result :)**

In [ ]:
fraction_movies[['title', 'vote_count', 'vote_average', 'score']].head(20)

title  vote_count  vote_average     score
314           The Shawshank Redemption      8358.0           8.5  8.445869
834                      The Godfather      6024.0           8.5  8.425439
10309      Dilwale Dulhania Le Jayenge       661.0           9.1  8.421453
12481                  The Dark Knight     12269.0           8.3  8.265477
2843                        Fight Club      9678.0           8.3  8.256385
292                       Pulp Fiction      8670.0           8.3  8.251406
522                   Schindler's List      4436.0           8.3  8.206639
23673                         Whiplash      4376.0           8.3  8.205404
5481                     Spirited Away      3968.0           8.3  8.196055
2211                 Life Is Beautiful      3643.0           8.3  8.187171
1178            The Godfather: Part II      3418.0           8.3  8.180076
1152   One Flew Over the Cuckoo's Nest      3001.0           8.3  8.164256
351                       Forrest Gump      8147.0           8.2  8.150272
1154           The Empire Strikes Back      5998.0           8.2  8.132919
1176                            Psycho      2405.0           8.3  8.132715
18465                 The Intouchables      5410.0           8.2  8.125837
40251                       Your Name.      1030.0           8.5  8.112532
289             Leon: The Professional      4293.0           8.2  8.107234
3030                    The Green Mile      4166.0           8.2  8.104511
1170                        GoodFellas      3211.0           8.2  8.077459